File: claude/harness/steambench_harness/envs/game2048.py
"""2048 — a deterministic SteamBench arcade env.

Why it's a good benchmark game: trivial rules, enormous skill ceiling, and a
clean achievement ladder (reach-the-N-tile) whose designed rarities mirror a
real Steam game's distribution. Fully deterministic given ``(seed, actions)``:
all tile spawns come from ``self.rng``, so the server can replay-verify any run.
"""

from __future__ import annotations

from typing import Union

from ..protocol import AchievementSpec, ActionSpace, Env, Observation, VerifyMode
from ..registry import register

SIZE = 4
_DIRS = ["up", "down", "left", "right"]


def _slide_left(row: list[int]) -> tuple[list[int], int, bool]:
    """Slide+merge one row to the left. Returns (new_row, gained, changed)."""
    nonzero = [v for v in row if v]
    merged: list[int] = []
    gained = 0
    i = 0
    while i < len(nonzero):
        if i + 1 < len(nonzero) and nonzero[i] == nonzero[i + 1]:
            v = nonzero[i] * 2
            merged.append(v)
            gained += v
            i += 2
        else:
            merged.append(nonzero[i])
            i += 1
    merged += [0] * (len(row) - len(merged))
    return merged, gained, merged != row


@register
class Game2048(Env):
    env_id = "arcade/2048"
    appid = 9000001
    name = "2048"
    description = (
        "Slide numbered tiles on a 4x4 grid; equal tiles merge and double. "
        "Reach the highest tile you can before the board jams."
    )
    verify_mode = VerifyMode.REPLAY
    action_space = ActionSpace(_DIRS)
    achievements = [
        AchievementSpec("tile_64", "Getting Started", "Reach the 64 tile.", 0.80),
        AchievementSpec("tile_128", "Warmed Up", "Reach the 128 tile.", 0.62),
        AchievementSpec("tile_256", "Climbing", "Reach the 256 tile.", 0.40),
        AchievementSpec("tile_512", "Halfway There", "Reach the 512 tile.", 0.24),
        AchievementSpec("tile_1024", "Big Numbers", "Reach the 1024 tile.", 0.11),
        AchievementSpec("tile_2048", "2048!", "Reach the fabled 2048 tile.", 0.045),
        AchievementSpec("tile_4096", "Beyond", "Reach the 4096 tile.", 0.009),
        AchievementSpec("tile_8192", "Grandmaster", "Reach the 8192 tile.", 0.0009),
        AchievementSpec("score_10k", "High Roller", "Score 10,000 points.", 0.16),
        AchievementSpec("score_20k", "Point Hoarder", "Score 20,000 points.", 0.05),
    ]

    def __init__(self) -> None:
        super().__init__()
        self.board: list[list[int]] = [[0] * SIZE for _ in range(SIZE)]

    # ---- lifecycle -------------------------------------------------------- #

    def reset(self, seed: int = 0) -> Observation:
        self._begin(seed)
        self.board = [[0] * SIZE for _ in range(SIZE)]
        self._spawn()
        self._spawn()
        return self._observe(reward=0.0)

    def step(self, action: Union[int, str]) -> Observation:
        if self.done:
            return self._observe(reward=0.0)
        name = self.action_space.name(action)
        before_score = self.score
        moved = self._move(name)
        if moved:
            self._spawn()
        self.steps += 1
        self._check_achievements()
        if not self._has_moves():
            self.done = True
        return self._observe(reward=self.score - before_score)

    # ---- mechanics -------------------------------------------------------- #

    def _move(self, direction: str) -> bool:
        b = self.board
        if direction == "left":
            rows = b
            new, gained, changed = self._apply_rows(rows)
            self.board = new
        elif direction == "right":
            rows = [list(reversed(r)) for r in b]
            new, gained, changed = self._apply_rows(rows)
            self.board = [list(reversed(r)) for r in new]
        elif direction == "up":
            rows = self._transpose(b)
            new, gained, changed = self._apply_rows(rows)
            self.board = self._transpose(new)
        elif direction == "down":
            rows = [list(reversed(r)) for r in self._transpose(b)]
            new, gained, changed = self._apply_rows(rows)
            self.board = self._transpose([list(reversed(r)) for r in new])
        else:
            return False
        self.score += gained
        return changed

    def _apply_rows(self, rows: list[list[int]]) -> tuple[list[list[int]], int, bool]:
        out = []
        total_gained = 0
        changed = False
        for r in rows:
            nr, gained, ch = _slide_left(r)
            out.append(nr)
            total_gained += gained
            changed = changed or ch
        return out, total_gained, changed

    @staticmethod
    def _transpose(b: list[list[int]]) -> list[list[int]]:
        return [list(col) for col in zip(*b)]

    def _empty_cells(self) -> list[tuple[int, int]]:
        return [(r, c) for r in range(SIZE) for c in range(SIZE) if self.board[r][c] == 0]

    def _spawn(self) -> None:
        empties = self._empty_cells()
        if not empties:
            return
        r, c = self.rng.choice(empties)
        self.board[r][c] = 4 if self.rng.random() < 0.1 else 2

    def _has_moves(self) -> bool:
        if self._empty_cells():
            return True
        for r in range(SIZE):
            for c in range(SIZE):
                v = self.board[r][c]
                if c + 1 < SIZE and self.board[r][c + 1] == v:
                    return True
                if r + 1 < SIZE and self.board[r + 1][c] == v:
                    return True
        return False

    def legal_actions(self) -> list[str]:
        legal = []
        snapshot = [row[:] for row in self.board]
        saved_score = self.score
        for d in _DIRS:
            if self._move(d):
                legal.append(d)
            self.board = [row[:] for row in snapshot]
            self.score = saved_score
        return legal

    @property
    def max_tile(self) -> int:
        return max(max(row) for row in self.board)

    def _check_achievements(self) -> None:
        mt = self.max_tile
        for n, aid in [
            (64, "tile_64"),
            (128, "tile_128"),
            (256, "tile_256"),
            (512, "tile_512"),
            (1024, "tile_1024"),
            (2048, "tile_2048"),
            (4096, "tile_4096"),
            (8192, "tile_8192"),
        ]:
            if mt >= n:
                self._unlock(aid)
        if self.score >= 10_000:
            self._unlock("score_10k")
        if self.score >= 20_000:
            self._unlock("score_20k")

    # ---- rendering -------------------------------------------------------- #

    def render(self) -> str:
        width = max(5, len(str(self.max_tile)) + 1)
        lines = []
        for row in self.board:
            lines.append("".join((str(v) if v else ".").center(width) for v in row))
        return f"score={self.score} max={self.max_tile}\n" + "\n".join(lines)

    def _observe(self, reward: float) -> Observation:
        legal = self.legal_actions()
        return Observation(
            step=self.steps,
            state={
                "board": [row[:] for row in self.board],
                "score": self.score,
                "max_tile": self.max_tile,
                "moves": self.steps,
            },
            text=self.render(),
            legal_actions=legal,
            score=float(self.score),
            done=self.done,
            reward=float(reward),
            info={"max_tile": self.max_tile, "newly": []},
        )

File: claude/web/src/lib/arcade/game2048.ts
/**
 * 2048 — a deterministic SteamBench arcade env.
 *
 * TypeScript mirror of `harness/steambench_harness/envs/game2048.py`. Fully
 * deterministic given `(seed, actions)`: all tile spawns come from `this.rng`,
 * so the server can replay-verify any run.
 */

import { ActionSpace, Env } from "./base";
import type { AchievementSpec, Action, Observation } from "./types";

const SIZE = 4;
const DIRS = ["up", "down", "left", "right"] as const;

/** Slide+merge one row to the left. Returns [newRow, gained, changed]. */
export function slideLeft(row: number[]): [number[], number, boolean] {
  const nonzero = row.filter((v) => v);
  const merged: number[] = [];
  let gained = 0;
  let i = 0;
  while (i < nonzero.length) {
    if (i + 1 < nonzero.length && nonzero[i] === nonzero[i + 1]) {
      const v = nonzero[i] * 2;
      merged.push(v);
      gained += v;
      i += 2;
    } else {
      merged.push(nonzero[i]);
      i += 1;
    }
  }
  while (merged.length < row.length) {
    merged.push(0);
  }
  const changed = !arraysEqual(merged, row);
  return [merged, gained, changed];
}

function arraysEqual(a: number[], b: number[]): boolean {
  if (a.length !== b.length) return false;
  for (let i = 0; i < a.length; i++) {
    if (a[i] !== b[i]) return false;
  }
  return true;
}

/** Mirror of Python `str(...).center(width)` (extra pad to the right). */
function center(s: string, width: number): string {
  if (s.length >= width) return s;
  const total = width - s.length;
  const left = Math.floor(total / 2);
  const right = total - left;
  return " ".repeat(left) + s + " ".repeat(right);
}

export class Game2048 extends Env {
  static env_id = "arcade/2048";
  static appid = 9000001;
  static displayName = "2048";
  static description =
    "Slide numbered tiles on a 4x4 grid; equal tiles merge and double. " +
    "Reach the highest tile you can before the board jams.";
  static verify_mode = "replay" as const;
  static action_space = new ActionSpace([...DIRS]);
  static achievements: AchievementSpec[] = [
    { id: "tile_64", name: "Getting Started", description: "Reach the 64 tile.", rarity_hint: 0.8 },
    { id: "tile_128", name: "Warmed Up", description: "Reach the 128 tile.", rarity_hint: 0.62 },
    { id: "tile_256", name: "Climbing", description: "Reach the 256 tile.", rarity_hint: 0.4 },
    { id: "tile_512", name: "Halfway There", description: "Reach the 512 tile.", rarity_hint: 0.24 },
    { id: "tile_1024", name: "Big Numbers", description: "Reach the 1024 tile.", rarity_hint: 0.11 },
    { id: "tile_2048", name: "2048!", description: "Reach the fabled 2048 tile.", rarity_hint: 0.045 },
    { id: "tile_4096", name: "Beyond", description: "Reach the 4096 tile.", rarity_hint: 0.009 },
    { id: "tile_8192", name: "Grandmaster", description: "Reach the 8192 tile.", rarity_hint: 0.0009 },
    { id: "score_10k", name: "High Roller", description: "Score 10,000 points.", rarity_hint: 0.16 },
    { id: "score_20k", name: "Point Hoarder", description: "Score 20,000 points.", rarity_hint: 0.05 },
  ];

  board: number[][];

  constructor() {
    super();
    this.board = Game2048.emptyBoard();
  }

  private static emptyBoard(): number[][] {
    const b: number[][] = [];
    for (let r = 0; r < SIZE; r++) {
      b.push(new Array<number>(SIZE).fill(0));
    }
    return b;
  }

  // ---- lifecycle --------------------------------------------------------- //

  reset(seed = 0): Observation {
    this.begin(seed);
    this.board = Game2048.emptyBoard();
    this.spawn();
    this.spawn();
    return this.observe(0.0);
  }

  step(action: Action): Observation {
    if (this.done) {
      return this.observe(0.0);
    }
    const name = this.actionSpace.name(action);
    const beforeScore = this.score;
    const moved = this.move(name);
    if (moved) {
      this.spawn();
    }
    this.steps += 1;
    this.checkAchievements();
    if (!this.hasMoves()) {
      this.done = true;
    }
    return this.observe(this.score - beforeScore);
  }

  // ---- mechanics --------------------------------------------------------- //

  private move(direction: string): boolean {
    const b = this.board;
    let changed = false;
    let gained = 0;
    if (direction === "left") {
      const [next, g, ch] = this.applyRows(b);
      this.board = next;
      gained = g;
      changed = ch;
    } else if (direction === "right") {
      const rows = b.map((r) => [...r].reverse());
      const [next, g, ch] = this.applyRows(rows);
      this.board = next.map((r) => [...r].reverse());
      gained = g;
      changed = ch;
    } else if (direction === "up") {
      const rows = Game2048.transpose(b);
      const [next, g, ch] = this.applyRows(rows);
      this.board = Game2048.transpose(next);
      gained = g;
      changed = ch;
    } else if (direction === "down") {
      const rows = Game2048.transpose(b).map((r) => [...r].reverse());
      const [next, g, ch] = this.applyRows(rows);
      this.board = Game2048.transpose(next.map((r) => [...r].reverse()));
      gained = g;
      changed = ch;
    } else {
      return false;
    }
    this.score += gained;
    return changed;
  }

  private applyRows(rows: number[][]): [number[][], number, boolean] {
    const out: number[][] = [];
    let totalGained = 0;
    let changed = false;
    for (const r of rows) {
      const [nr, gained, ch] = slideLeft(r);
      out.push(nr);
      totalGained += gained;
      changed = changed || ch;
    }
    return [out, totalGained, changed];
  }

  private static transpose(b: number[][]): number[][] {
    const out: number[][] = [];
    for (let c = 0; c < b[0].length; c++) {
      const col: number[] = [];
      for (let r = 0; r < b.length; r++) {
        col.push(b[r][c]);
      }
      out.push(col);
    }
    return out;
  }

  private emptyCells(): [number, number][] {
    const cells: [number, number][] = [];
    for (let r = 0; r < SIZE; r++) {
      for (let c = 0; c < SIZE; c++) {
        if (this.board[r][c] === 0) {
          cells.push([r, c]);
        }
      }
    }
    return cells;
  }

  private spawn(): void {
    const empties = this.emptyCells();
    if (empties.length === 0) {
      return;
    }
    // CRITICAL ORDER: pick the cell first, THEN draw the value.
    const [r, c] = this.rng.choice(empties);
    this.board[r][c] = this.rng.random() < 0.1 ? 4 : 2;
  }

  private hasMoves(): boolean {
    if (this.emptyCells().length > 0) {
      return true;
    }
    for (let r = 0; r < SIZE; r++) {
      for (let c = 0; c < SIZE; c++) {
        const v = this.board[r][c];
        if (c + 1 < SIZE && this.board[r][c + 1] === v) {
          return true;
        }
        if (r + 1 < SIZE && this.board[r + 1][c] === v) {
          return true;
        }
      }
    }
    return false;
  }

  legalActions(): string[] {
    const legal: string[] = [];
    const snapshot = this.board.map((row) => [...row]);
    const savedScore = this.score;
    for (const d of DIRS) {
      if (this.move(d)) {
        legal.push(d);
      }
      this.board = snapshot.map((row) => [...row]);
      this.score = savedScore;
    }
    return legal;
  }

  get maxTile(): number {
    let m = 0;
    for (const row of this.board) {
      for (const v of row) {
        if (v > m) m = v;
      }
    }
    return m;
  }

  private checkAchievements(): void {
    const mt = this.maxTile;
    const ladder: [number, string][] = [
      [64, "tile_64"],
      [128, "tile_128"],
      [256, "tile_256"],
      [512, "tile_512"],
      [1024, "tile_1024"],
      [2048, "tile_2048"],
      [4096, "tile_4096"],
      [8192, "tile_8192"],
    ];
    for (const [n, aid] of ladder) {
      if (mt >= n) {
        this.unlock(aid);
      }
    }
    if (this.score >= 10000) {
      this.unlock("score_10k");
    }
    if (this.score >= 20000) {
      this.unlock("score_20k");
    }
  }

  // ---- rendering --------------------------------------------------------- //

  render(): string {
    const width = Math.max(5, String(this.maxTile).length + 1);
    const lines: string[] = [];
    for (const row of this.board) {
      lines.push(row.map((v) => center(v ? String(v) : ".", width)).join(""));
    }
    return `score=${this.score} max=${this.maxTile}\n` + lines.join("\n");
  }

  private observe(reward: number): Observation {
    const legal = this.legalActions();
    return {
      step: this.steps,
      state: {
        board: this.board.map((row) => [...row]),
        score: this.score,
        max_tile: this.maxTile,
        moves: this.steps,
      },
      text: this.render(),
      frame: null,
      legal_actions: legal,
      score: this.score,
      done: this.done,
      reward: reward,
      info: { max_tile: this.maxTile, newly: [] as string[] },
    };
  }
}
